lib/Basic/ParseableOutput.cpp

```diff
@@ -0,0 +1,484 @@
+      : CommandBasedMessage(Kind, Cmd) {
+    Executable = Cmd.getExecutable();
+    for (const auto &A : Cmd.getArguments()) {
+      Arguments.push_back(A);
+    }
+    llvm::raw_string_ostream wrapper(CommandLine);
+    Cmd.printCommandLine(wrapper, "");
+    wrapper.flush();
+
+    for (const Action *A : Cmd.getSource().getInputs()) {
+      if (const auto *IA = dyn_cast<InputAction>(A))
+        Inputs.push_back(CommandInput(IA->getInputArg().getValue()));
+    }
+
+    for (const driver::Job *J : Cmd.getInputs()) {
+      auto OutFiles = J->getOutput().getPrimaryOutputFilenames();
+      if (const auto *BJAction = dyn_cast<BackendJobAction>(&Cmd.getSource())) {
+        Inputs.push_back(CommandInput(OutFiles[BJAction->getInputIndex()]));
+      } else {
+        for (llvm::StringRef FileName : OutFiles) {
+          Inputs.push_back(CommandInput(FileName));
+        }
+      }
+    }
+
+    // TODO: set up Outputs appropriately.
+    file_types::ID PrimaryOutputType = Cmd.getOutput().getPrimaryOutputType();
+    if (PrimaryOutputType != file_types::TY_Nothing) {
+      for (llvm::StringRef OutputFileName :
+           Cmd.getOutput().getPrimaryOutputFilenames()) {
+        Outputs.push_back(OutputPair(PrimaryOutputType, OutputFileName.str()));
+      }
+    }
+    file_types::forAllTypes([&](file_types::ID Ty) {
+      for (auto Output : Cmd.getOutput().getAdditionalOutputsForType(Ty)) {
+        Outputs.push_back(OutputPair(Ty, Output.str()));
+      }
+    });
+  }
+
+  void provideMapping(swift::json::Output &out) override {
+    Message::provideMapping(out);
+    out.mapRequired("command", CommandLine); // Deprecated, do not document
+    out.mapRequired("command_executable", Executable);
+    out.mapRequired("command_arguments", Arguments);
+    out.mapOptional("inputs", Inputs);
+    out.mapOptional("outputs", Outputs);
+  }
+};
+
+class DetailedInvocationBasedMessage : public InvocationBasedMessage {
+  ArrayRef<const char *> Args;
+  std::string Executable;
+  SmallVector<std::string, 16> Arguments;
+  std::string CommandLine;
+  SmallVector<CommandInput, 4> Inputs;
+  SmallVector<OutputPair, 8> Outputs;
+
+public:
+  DetailedInvocationBasedMessage(StringRef Kind,
+                                 const CompilerInvocation &Invocation,
+                                 const InputFile &PrimaryInput,
+                                 ArrayRef<const char *> Args)
+      : InvocationBasedMessage(Kind, Invocation), Args(Args) {
+    // Command line and arguments
+    Executable = Invocation.getFrontendOptions().MainExecutablePath;
+    CommandLine += Executable;
+    for (const auto &A : Args) {
+      Arguments.push_back(A);
+      CommandLine += std::string(" ") + A;
+    }
+
+    // Primary Input only
+    Inputs.push_back(CommandInput(PrimaryInput.getFileName()));
+
+    // Output for this Primary
+    auto OutputFile = PrimaryInput.outputFilename();
+    Outputs.push_back(OutputPair(file_types::lookupTypeForExtension(
```

Comment:
Like below, I think you could avoid using `lookupTypeForExtension()` here. Would `FrontendOptions::formatForPrincipalOutputFileForAction()` get the right file type?